### PR TITLE
Add maxSize downsampling option to APNGImage

### PR DESCRIPTION
**Pull request by plateaukao**
## Motivation

Large APNGs render at full native resolution today — there is no downsampling or max-dimension guard. A single 10240×7680 image allocates a full-resolution RGBA canvas (~300 MB) plus the output/previous `CGImage`s per displayed image (~600 MB+ in steady state). A few such images in a list (e.g. a `UICollectionView`) exceed the per-app memory limit on iPhone and the app gets jetsam-killed.

## What this does

Adds an optional `maxSize: CGSize?` to every `APNGImage` initializer. When set, the decoder derives a `renderScale = min(1, fit)` and the **entire compositing pipeline runs in that scaled space** — the canvas buffer, every cached frame, and all frame rects shrink to fit `maxSize`, which bounds memory.

```swift
// Caps a huge image to ~screen size: renders at 1024×768 instead of 10240×7680
let image = try APNGImage(named: "huge", maxSize: CGSize(width: 1024, height: 768))
imageView.image = image
```

Key points:

- **Compositing in scaled space**: the `CGContext` is allocated at `renderWidth × renderHeight`, and the dispose region, `.previous` crop, and blend draw rects all pass through a single `CGAffineTransform(scaleX:y:)` so frame boundaries stay coincident.
- **Free downsampling**: frames still decode natively, then `CGContext.draw(_, in: scaledRect)` resamples them as they're composited.
- **No behavior change by default**: `maxSize: nil` keeps `renderScale = 1.0` and produces byte-identical geometry to before.
- **Upper bound only**: an image already smaller than `maxSize` is left at native size (never upscaled).
- **Cache estimate** uses the scaled dimensions, so a downsampled loop that now fits under `maximumCacheSize` can regain caching.

## Caveat

Fractional frame offsets in partial-update APNGs can show minor edge-resampling artifacts when scaled; full-frame APNGs (the common case) downsample cleanly.

## Tests

- New tests in `APNGImageRendererTests`: `testRenderDownsamplesToMaxSize` (asserts a 100×100 image → 50×50 across multiple composited frames, plus a native-path regression guard) and `testMaxSizeLargerThanNativeDoesNotUpscale`.
- `swift build` clean; **all 87 tests pass** (85 existing + 2 new).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: Source/APNGKit/APNGDecoder.swift [hunk 1]
```diff
@@ -46,8 +51,31 @@ class APNGDecoder {
...
+    // The height of the rendering canvas in pixels, after applying `renderScale`.
+    var renderHeight: Int { scaledLength(imageHeader.height) }
+    // The bytes per row of the rendering canvas, after applying `renderScale`.
+    var renderBytesPerRow: Int { renderWidth * Int(imageHeader.bytesPerPixel) }
+
+    // Scales a single length (a width or a height) from the native coordinate space into the render space, clamping to
+    // a minimum of one pixel so the canvas is never degenerate.
+    private func scaledLength(_ value: Int) -> Int {
+        guard renderScale < 1.0 else { return value }
+        return max(1, Int((CGFloat(value) * renderScale).rounded()))
```
Comment (onevcat) [inline Source/APNGKit/APNGDecoder.swift]: Although not likely to be a real case, when `value == 0`, this method returns `1` as its result, which is not good.

File: Source/APNGKit/APNGDecoder.swift [hunk 2]
```diff
@@ -119,7 +161,16 @@ class APNGDecoder {
         } else { // Optimization: Auto determine if we want to cache the image based on image information.
             if acTLResult.chunk.numberOfPlays == 0 {
                 // Although it is not accurate enough, we only use the image header and animation control chunk to estimate.
-                let estimatedTotalBytes = imageHeader.height * imageHeader.bytesPerRow * numberOfFrames
+                // Use the render (downsampled) dimensions, since that is the size each cached frame actually takes. These
+                // are computed inline rather than via `renderWidth`/`renderHeight` because `self` is not yet fully
+                // initialized here.
+                let scaledWidth = renderScale < 1.0
+                    ? max(1, Int((CGFloat(imageHeader.width) * renderScale).rounded()))
```
Comment (onevcat) [inline Source/APNGKit/APNGDecoder.swift]: Maybe we should use the `scaledLength` just defined above.

File: Source/APNGKit/APNGDecoder.swift [hunk 3]
```diff
@@ -119,7 +161,16 @@ class APNGDecoder {
...
+                    : imageHeader.width
+                let scaledHeight = renderScale < 1.0
+                    ? max(1, Int((CGFloat(imageHeader.height) * renderScale).rounded()))
```
Comment (onevcat) [inline Source/APNGKit/APNGDecoder.swift]: ditto

File: Source/APNGKit/APNGImageRenderer.swift [hunk 4]
```diff
@@ -449,13 +453,15 @@ extension APNGImageRenderer {
             }
         }
         
-        // Blend & Draw the new frame
+        // Blend & Draw the new frame. The frame's destination rectangle is scaled into render space; drawing the
+        // natively-decoded `nextFrameImage` into a smaller rectangle lets Core Graphics downsample it for us.
+        let frameRenderRect = decoder.renderRect(frame.normalizedRect(fullHeight: decoder.imageHeader.height))
```
Comment (onevcat) [inline Source/APNGKit/APNGImageRenderer.swift:462]: Nice & clean. Thank you.

File: Source/APNGKit/APNGImageRenderer.swift [hunk 5]
```diff
@@ -436,7 +439,8 @@ extension APNGImageRenderer {
                 previousOutputImage = outputBuffer.makeImage()
             case .previous:
                 if let previousOutputImage = previousOutputImage {
-                    if let cropped = previousOutputImage.cropping(to: displayingFrame.frameControl.cgRect) {
+                    // `previousOutputImage` is already at render scale, so crop it in render space too.
+                    if let cropped = previousOutputImage.cropping(to: decoder.renderRect(displayingFrame.frameControl.cgRect)) {
                         outputBuffer.clear(displayingRegion)
                         outputBuffer.draw(cropped, in: displayingRegion)
```
[the remark this answers is not in the document]
Comment (plateaukao) [reply Source/APNGKit/APNGImageRenderer.swift]: Good catch — fixed in `ff93124`. The `.previous` crop rect is now integralized and clamped to the image bounds before `cropping(to:)`:

```swift
let imageBounds = CGRect(x: 0, y: 0, width: previousOutputImage.width, height: previousOutputImage.height)
let cropRect = decoder.renderRect(displayingFrame.frameControl.cgRect).integral.intersection(imageBounds)
if let cropped = previousOutputImage.cropping(to: cropRect) { ... }
```

so a fractional or slightly-out-of-bounds scaled rect can no longer make it return `nil` and break disposal.

File: Source/APNGKit/APNGDecoder.swift (hunk 1 above)
Comment (plateaukao) [reply Source/APNGKit/APNGDecoder.swift]: Fixed in `ff93124`. A zero length now stays zero — only positive lengths are clamped to a minimum of one pixel:

```swift
static func scaledLength(_ value: Int, scale: CGFloat) -> Int {
    guard scale < 1.0, value > 0 else { return value }
    return max(1, Int((CGFloat(value) * scale).rounded()))
}
```

Also added a `testScaledLength` unit test covering the zero / clamp / rounding cases.

File: Source/APNGKit/APNGDecoder.swift (hunk 2 above)
Comment (plateaukao) [reply Source/APNGKit/APNGDecoder.swift]: Done in `ff93124`. I made `scaledLength` a `static` method so it can be called during `init` (before `self` is fully initialized), and the cache estimate now reuses it instead of the duplicated arithmetic:

```swift
let scaledWidth = Self.scaledLength(imageHeader.width, scale: renderScale)
let scaledHeight = Self.scaledLength(imageHeader.height, scale: renderScale)
```

File: Source/APNGKit/APNGDecoder.swift (hunk 3 above)
Comment (plateaukao) [reply Source/APNGKit/APNGDecoder.swift]: Same fix as above — the estimate now uses `Self.scaledLength(...)` for both width and height in `ff93124`.

File: Source/APNGKit/APNGImageRenderer.swift (hunk 4 above)
Comment (plateaukao) [reply Source/APNGKit/APNGImageRenderer.swift:462]: 🙏 Thanks!

File: Source/APNGKit/APNGImage.swift [hunk 6]
```diff
@@ -136,15 +136,18 @@ public class APNGImage {
     /// - Parameters:
     ///   - name: The name of the image file in the main bundle.
     ///   - decodingOptions: The decoding options being used while decoding the image data.
+    ///   - maxSize: The maximum pixel size the image should be rendered at. If the image is larger than this size, it is
+    ///     downsampled to fit, which bounds its memory footprint. Pass `nil` (the default) to render at native size.
```
[the remark this answers is not in the document]
Comment (plateaukao) [reply Source/APNGKit/APNGImage.swift]: Clarified the docs rather than changing `size`, in `ff93124`. Keeping `size` (and `APNGImageView.intrinsicContentSize`) at the native logical dimensions is intentional: `maxSize` is a memory / rendering-resolution control, not a layout change — the downsampled bitmap is scaled up to fill the same logical size when displayed. The doc block now states this explicitly:

> This affects only the rendering resolution and memory use: the image's logical `size` (and an `APNGImageView`'s `intrinsicContentSize`) still reflects the native dimensions, and the downsampled output is scaled up to fill that size when displayed.

File: Source/APNGKit/APNGImage.swift [hunk 7]
```diff
@@ -153,6 +156,8 @@ public class APNGImage {
     ///   - decodingOptions: The decoding options being used while decoding the image data.
     ///   - bundle: The bundle in which APNGKit should search in for the image.
     ///   - subpath: The subdirectory path in the bundle where the image is put.
+    ///   - maxSize: The maximum pixel size the image should be rendered at. If the image is larger than this size, it is
+    ///     downsampled to fit, which bounds its memory footprint. Pass `nil` (the default) to render at native size.
```
[the remark this answers is not in the document]
Comment (plateaukao) [reply Source/APNGKit/APNGImage.swift]: Clarified the docs rather than changing `size`, in `ff93124`. Keeping `size` (and `APNGImageView.intrinsicContentSize`) at the native logical dimensions is intentional: `maxSize` is a memory / rendering-resolution control, not a layout change — the downsampled bitmap is scaled up to fill the same logical size when displayed. The doc block now states this explicitly:

> This affects only the rendering resolution and memory use: the image's logical `size` (and an `APNGImageView`'s `intrinsicContentSize`) still reflects the native dimensions, and the downsampled output is scaled up to fill that size when displayed.

File: Source/APNGKit/APNGImage.swift [hunk 8]
```diff
@@ -161,44 +166,51 @@ public class APNGImage {
...
+        try self.init(fileURL: resource.fileURL, scale: resource.scale, decodingOptions: decodingOptions, maxSize: maxSize)
     }
     
     /// Creates an APNG image object using the file path.
     /// - Parameters:
     ///   - filePath: The path of APNG file.
     ///   - scale: The desired image scale. If not set, APNGKit will guess from the file name.
     ///   - decodingOptions: The decoding options being used while decoding the image data.
+    ///   - maxSize: The maximum pixel size the image should be rendered at. If the image is larger than this size, it is
+    ///     downsampled to fit, which bounds its memory footprint. Pass `nil` (the default) to render at native size.
```
[the remark this answers is not in the document]
Comment (plateaukao) [reply Source/APNGKit/APNGImage.swift]: Clarified the docs rather than changing `size`, in `ff93124`. Keeping `size` (and `APNGImageView.intrinsicContentSize`) at the native logical dimensions is intentional: `maxSize` is a memory / rendering-resolution control, not a layout change — the downsampled bitmap is scaled up to fill the same logical size when displayed. The doc block now states this explicitly:

> This affects only the rendering resolution and memory use: the image's logical `size` (and an `APNGImageView`'s `intrinsicContentSize`) still reflects the native dimensions, and the downsampled output is scaled up to fill that size when displayed.

File: Source/APNGKit/APNGImage.swift [hunk 9]
```diff
@@ -161,44 +166,51 @@ public class APNGImage {
...
+        try self.init(fileURL: fileURL, scale: scale, decodingOptions: decodingOptions, maxSize: maxSize)
     }
     
     /// Creates an APNG image object using the file URL.
     /// - Parameters:
     ///   - fileURL: The URL of APNG file on disk.
     ///   - scale: The desired image scale. If not set, APNGKit will guess from the file name.
     ///   - decodingOptions: The decoding options being used while decoding the image data.
+    ///   - maxSize: The maximum pixel size the image should be rendered at. If the image is larger than this size, it is
+    ///     downsampled to fit, which bounds its memory footprint. Pass `nil` (the default) to render at native size.
```
[the remark this answers is not in the document]
Comment (plateaukao) [reply Source/APNGKit/APNGImage.swift]: Clarified the docs rather than changing `size`, in `ff93124`. Keeping `size` (and `APNGImageView.intrinsicContentSize`) at the native logical dimensions is intentional: `maxSize` is a memory / rendering-resolution control, not a layout change — the downsampled bitmap is scaled up to fill the same logical size when displayed. The doc block now states this explicitly:

> This affects only the rendering resolution and memory use: the image's logical `size` (and an `APNGImageView`'s `intrinsicContentSize`) still reflects the native dimensions, and the downsampled output is scaled up to fill that size when displayed.

File: Source/APNGKit/APNGImage.swift [hunk 10]
```diff
@@ -218,15 +230,18 @@ public class APNGImage {
     ///   - data: The data containing APNG information and frames.
     ///   - scale: The desired image scale. If not set, `1.0` is used.
     ///   - decodingOptions: The decoding options being used while decoding the image data.
+    ///   - maxSize: The maximum pixel size the image should be rendered at. If the image is larger than this size, it is
+    ///     downsampled to fit, which bounds its memory footprint. Pass `nil` (the default) to render at native size.
```
[the remark this answers is not in the document]
Comment (plateaukao) [reply Source/APNGKit/APNGImage.swift]: Clarified the docs rather than changing `size`, in `ff93124`. Keeping `size` (and `APNGImageView.intrinsicContentSize`) at the native logical dimensions is intentional: `maxSize` is a memory / rendering-resolution control, not a layout change — the downsampled bitmap is scaled up to fill the same logical size when displayed. The doc block now states this explicitly:

> This affects only the rendering resolution and memory use: the image's logical `size` (and an `APNGImageView`'s `intrinsicContentSize`) still reflects the native dimensions, and the downsampled output is scaled up to fill that size when displayed.
